beacon-chain/rpc/beacon_server.go

```diff
@@ -97,24 +99,117 @@ func (bs *BeaconServer) LatestAttestation(req *ptypes.Empty, stream pb.BeaconSer
-	return &pb.Eth1DataResponse{}, nil
+	beaconState, err := bs.beaconDB.State()
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch beacon state: %v", err)
+	}
+	dataVoteObjects := []*pbp2p.Eth1DataVote{}
+	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
+	for _, vote := range beaconState.Eth1DataVotes {
+		eth1Hash := bytesutil.ToBytes32(vote.Eth1Data.BlockHash32)
+		// Verify the block from the vote's block hash exists in the eth1.0 chain and fetch its height.
+		blockExists, blockNumber, err := bs.powChainService.BlockExists(eth1Hash)
+		if err != nil {
+			log.Errorf("Could not verify block with hash exists in Eth1 chain: %#x: %v", eth1Hash, err)
+			continue
+		}
+		// Fetch the current canonical chain height from the eth1.0 chain.
+		currentHeight := bs.powChainService.LatestBlockNumber()
+		// Fetch the height of the block pointed to by the beacon state's latest_eth1_data.block_hash
+		// in the canonical, eth1.0 chain.
+		stateLatestEth1Hash := bytesutil.ToBytes32(beaconState.LatestEth1Data.BlockHash32)
+		_, stateLatestEth1Height, err := bs.powChainService.BlockExists(stateLatestEth1Hash)
+		if err != nil {
+			log.Errorf("Could not verify block with hash exists in Eth1 chain: %#x: %v", eth1Hash, err)
+			continue
+		}
+		// Let dataVoteObjects be the set of Eth1DataVote objects vote in state.eth1_data_votes where:
+		// vote.eth1_data.block_hash is the hash of an eth1.0 block that is:
+		//   (i) part of the canonical chain
+		//   (ii) >= ETH1_FOLLOW_DISTANCE blocks behind the head
+		//   (iii) newer than state.latest_eth1_data.block_data.
+		// vote.eth1_data.deposit_root is the deposit root of the eth1.0 deposit contract
+		// at the block defined by vote.eth1_data.block_hash.
+		isBehindFollowDistance := blockNumber.Add(blockNumber, big.NewInt(eth1FollowDistance)).Cmp(currentHeight) == -1
+		isAheadStateLatestEth1Data := blockNumber.Cmp(stateLatestEth1Height) == 1
+		if blockExists && isBehindFollowDistance && isAheadStateLatestEth1Data {
+			dataVoteObjects = append(dataVoteObjects, vote)
+		}
+	}
+
+	// Now we handle the following two scenarios:
+	// If dataVoteObjects is empty:
+	// Let block_hash be the block hash of the ETH1_FOLLOW_DISTANCE'th ancestor of the head of
+	// the canonical eth1.0 chain.
+	// Let deposit_root be the deposit root of the eth1.0 deposit contract in the
+	// post-state of the block referenced by block_hash.
+	if len(dataVoteObjects) == 0 {
+		// Fetch the current canonical chain height from the eth1.0 chain.
+		currentHeight := bs.powChainService.LatestBlockNumber()
+		ancestorHeight := currentHeight.Sub(currentHeight, big.NewInt(eth1FollowDistance))
+		blockHash, err := bs.powChainService.BlockHashByHeight(ancestorHeight)
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch ETH1_FOLLOW_DISTANCE ancestor: %v", err)
+		}
+		depositRoot := bs.powChainService.DepositRoot()
```

Comment:
I don't think this is correct. We should get the `depositRoot` that is referenced by `blockHash` from `ancestorHeight`.

bs.powChainService.DepositRoot() just returns you the latest deposit root right?

Reply:
@terenc3t there is no way to actually do that now, we dont keep a list of all the previous deposit roots. When we get a deposit log we just update the deposit trie and create a new root. Also deposit root is dependant on the merkle index of the deposit not the block number. We could have multiple deposits in a block , generating multiple deposit roots

Reply:
that's fine we can't do this now but where is the comment or note saying our implementation does diverge from the spec